hail/src/main/scala/is/hail/expr/ir/ArraySorter.scala

```diff
@@ -2,75 +2,207 @@ package is.hail.expr.ir
 
 import is.hail.annotations.Region
 import is.hail.asm4s._
-import is.hail.types.physical.{PCanonicalArray, PCanonicalDict, PCanonicalSet, PCode, PIndexableCode, PType, typeToTypeInfo}
+import is.hail.types.physical.{PCanonicalArray, PCanonicalDict, PCanonicalSet, PIndexableCode}
+import is.hail.types.virtual.{TArray, TDict, TSet, Type}
+import is.hail.utils.FastIndexedSeq
+
+import scala.language.existentials
 
 class ArraySorter(r: EmitRegion, array: StagedArrayBuilder) {
-  val typ: PType = array.elt
-  val ti: TypeInfo[_] = typeToTypeInfo(typ)
+  val ti: TypeInfo[_] = array.elt.ti
   val mb: EmitMethodBuilder[_] = r.mb
 
-  def sort(sorter: DependentEmitFunctionBuilder[_]): Code[Unit] = {
-    val localF = ti match {
-      case BooleanInfo => mb.genFieldThisRef[AsmFunction2[Boolean, Boolean, Boolean]]()
-      case IntInfo => mb.genFieldThisRef[AsmFunction2[Int, Int, Boolean]]()
-      case LongInfo => mb.genFieldThisRef[AsmFunction2[Int, Int, Boolean]]()
-      case FloatInfo => mb.genFieldThisRef[AsmFunction2[Long, Long, Boolean]]()
-      case DoubleInfo => mb.genFieldThisRef[AsmFunction2[Double, Double, Boolean]]()
+  private[this] var prunedMissing: Boolean = false
+
+  private[this] val workingArrayInfo = arrayInfo(array.ti)
+  private[this] val workingArray1 = mb.genFieldThisRef("sorter_working_array")(workingArrayInfo)
+  private[this] val workingArray2 = mb.genFieldThisRef("sorter_working_array")(workingArrayInfo)
+
+  private[this] def arrayRef(workingArray: Code[Array[_]]): UntypedCodeArray = new UntypedCodeArray(workingArray, array.ti)
+
+  def sort(cb: EmitCodeBuilder, region: Value[Region], comparesLessThan: (EmitCodeBuilder, Value[Region], Code[_], Code[_]) => Code[Boolean]): Unit = {
+
+    val sortMB = cb.emb.ecb.newEmitMethod("arraySorter_outer", FastIndexedSeq[ParamType](classInfo[Region]), UnitInfo)
```

Comment:
apparently this is leading to duplicate method errors.

Reply:
yeah this is supposed to be `genEmitMethod` which has the same signature but generates a name instead